windows/devices/uwb/UwbConnector.cxx

```diff
@@ -895,43 +868,256 @@ UwbConnector::NotificationListenerStop()
-    m_deviceEventCallbacks.push_back(token);
-    if (isFirstCallback) {
-        NotificationListenerStart();
+    auto callbackWeak = accessor(callbacks);
+    auto callbackShared = callbackWeak.lock();
+    if (not callbackShared) {
+        return std::shared_ptr<::uwb::RegisteredDeviceCallbackToken>();
+    }
+    auto callback = *callbackShared;
+    if (not callback) {
+        return std::shared_ptr<::uwb::RegisteredDeviceCallbackToken>();
     }
-    return token;
+    return tokenizer(callbackWeak);
 }
 
-std::weak_ptr<::uwb::RegisteredCallbackToken>
-UwbConnector::RegisterSessionEventCallbacks(uint32_t sessionId, std::weak_ptr<::uwb::UwbRegisteredSessionEventCallbacks> callbacks)
+::uwb::UwbRegisteredDeviceEventCallbackTokens
+UwbConnector::RegisterDeviceEventCallbacks(::uwb::UwbRegisteredDeviceEventCallbacks callbacks)
 {
     std::lock_guard eventCallbacksLockExclusive{ m_eventCallbacksGate };
     bool isFirstCallback = not CallbacksPresent();
-    auto token = std::make_shared<::uwb::RegisteredSessionCallbackToken>(sessionId, callbacks);
-    auto node = m_sessionEventCallbacks.extract(sessionId);
+
+    auto OnStatusChangedToken = GetToken<::uwb::UwbRegisteredDeviceEventCallbackTypes::OnStatusChanged>(
+        callbacks, [](auto&& cbs) {
+            return cbs.OnStatusChanged;
+        },
+        [this](auto&& cb) {
+            auto token = std::make_shared<::uwb::OnStatusChangedToken>(cb);
+            m_onStatusChangedCallbacks.push_back(token);
+            return token;
+        });
+
+    auto OnDeviceStatusChangedToken = GetToken<::uwb::UwbRegisteredDeviceEventCallbackTypes::OnDeviceStatusChanged>(
+        callbacks, [](auto&& cbs) {
+            return cbs.OnDeviceStatusChanged;
+        },
+        [this](auto&& cb) {
+            auto token = std::make_shared<::uwb::OnDeviceStatusChangedToken>(cb);
+            m_onDeviceStatusChangedCallbacks.push_back(token);
+            return token;
+        });
+
+    auto OnSessionStatusChangedToken = GetToken<::uwb::UwbRegisteredDeviceEventCallbackTypes::OnSessionStatusChanged>(
+        callbacks, [](auto&& cbs) {
+            return cbs.OnSessionStatusChanged;
+        },
+        [this](auto&& cb) {
+            auto token = std::make_shared<::uwb::OnSessionStatusChangedToken>(cb);
+            m_onSessionStatusChangedCallbacks.push_back(token);
+            return token;
+        });
+
+    return {
+        OnStatusChangedToken,
+        OnDeviceStatusChangedToken,
+        OnSessionStatusChangedToken
+    };
+}
+
+/**
+ * @brief Internal helper function to insert a token into the map of session tokens
+ *
+ * @tparam T the type of the token
+ * @param tokens the map of tokens to insert into
+ * @param sessionId
+ * @param token
+ */
+template <typename T>
+void
+InsertSessionToken(std::unordered_map<uint32_t, std::vector<std::shared_ptr<T>>>& tokens, uint32_t sessionId, std::shared_ptr<T> token)

```

Comment:
Could sessionId could be const here?

Reply:
Makes sense. I was thinking about it conceptually - I wouldn't expect the sessionId to ever need to change in this function. But I totally understand having non-`const` in function definitions.